osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/factories/Hls5Connector.java

```diff
@@ -97,24 +96,38 @@ public DlmsConnection connect(
     }
   }
 
+  /**
+   * Connect method without any error handling. Used by key recovery process to try to connect to a
+   * device with newly registered keys, without triggering a new key recovery process when the
+   * connection fails.
+   *
+   * @param device the device to connect to
+   * @param dlmsMessageListener a message listener (for tracing or debugging)
+   * @param keyProvider the method to use to get the required keys
+   * @return the created connection
+   * @throws IOException
+   * @throws OsgpException
+   * @see org.opensmartgridplatform.adapter.protocol.dlms.application.threads.RecoverKeyProcess
+   */
+  public DlmsConnection connectUnchecked(
+      final DlmsDevice device,
+      final DlmsMessageListener dlmsMessageListener,
+      final SecurityKeyProvider keyProvider)
+      throws IOException, OsgpException {
+    return this.createConnection(device, dlmsMessageListener, keyProvider);
+  }
+
   @Override
   protected void setSecurity(
-      final DlmsDevice device, final TcpConnectionBuilder tcpConnectionBuilder)
-      throws OsgpException {
+      final DlmsDevice device,
+      final SecurityKeyProvider provider,
+      final TcpConnectionBuilder tcpConnectionBuilder)
+      throws FunctionalException {
 
-    final String deviceIdentification = device.getDeviceIdentification();
-    final byte[] dlmsAuthenticationKey;
-    final byte[] dlmsEncryptionKey;
-    try {
-      final Map<SecurityKeyType, byte[]> encryptedKeys =
-          this.secretManagementService.getKeys(
-              deviceIdentification, Arrays.asList(E_METER_AUTHENTICATION, E_METER_ENCRYPTION));
-      dlmsAuthenticationKey = encryptedKeys.get(E_METER_AUTHENTICATION);
-      dlmsEncryptionKey = encryptedKeys.get(E_METER_ENCRYPTION);
-    } catch (final EncrypterException e) {
-      throw new FunctionalException(
-          FunctionalExceptionType.INVALID_DLMS_KEY_ENCRYPTION, ComponentType.PROTOCOL_DLMS, e);
-    }
+    final byte[] dlmsAuthenticationKey =
+        provider.getKey(device.getDeviceIdentification(), E_METER_AUTHENTICATION);
+    final byte[] dlmsEncryptionKey =
+        provider.getKey(device.getDeviceIdentification(), E_METER_ENCRYPTION);
```

Comment:
Based on something I commented on in another pull request recently, I think the code could be made a little more efficient by changing `SecurityKeyProvider` to match the signature of `getKeys`/`getNewKeys` instead of `getKey`/`getNewKey`. Then you could retrieve one or more keys in a single call to secret management, like in the code that has been replaced. Not sure how much of an undesirable effect the extra call via the newly introduced `SecurityKeyProvider` has.

Reply:
Replaced getKey with getKeys